Update contributing guide (#2382)
unitaryfund/mitiq: docs/CONTRIBUTING_DOCS.md

Documentation is found in `docs/source`, and is divided into the following sections:

-  a **guide**, whose content needs to be written from scratch for new features. For more details, go to [](contributing.md#proposing-a-new-feature-to-mitiq) and [](contributing_docs.md#adding-files-to-the-user-guide)
- **examples** which can be either a jupyter notebook or a MyST formatted notebook. A checklist is available in [](contributing_docs.md#adding-example-files)
- an **API-doc** part, which is (mostly) automatically generated. See [](contributing_docs.md#automatically-add-information-from-the-api-docs) for additional information.

Information in the docs should be added as markdown files using the MyST markdown syntax.

> If you are adding a new file (as opposed to editing an existing one), ensure to add it to an associated TOC so that it is discoverable. 

Reviewer: ```suggestion
If you are adding a new file (as opposed to editing an existing one), ensure to add it to an associated TOC so that it is discoverable.
```

We should set up a linter for this.

Author: If you use `make docs` without adding the new addition to the TOC, sphinx will complain that the file is not added anywhere.

Reviewer: Sorry, I mean we should set up a linter to catch white space at the end of lines.

Author: Ah. Will have to search for something. 

I see there's a sphinx extension for it but a quick glance shows it is for `.rst` files only: https://github.com/sphinx-contrib/sphinx-lint

To add information in the guide, please add markdown (`.md`) files to the `docs/guide` directory.
Remember to add new files to the guide's TOC file `docs/source/guide/guide.md`.

The different sections of a User's guide comprises of:

- How do I use `new_QEM_method` labeled as `new_QEM_method-1-intro.md`?
- When should I use `new_QEM_method` labeled as `new_QEM_method-2-use-case.md`?
- What additional options are available in `new_QEM_method` labeled as `new_QEM_method-3-options.md`?
- What happens when I use `new_QEM_method` labeled as `new_QEM_method-4-low-level.md`?
- What is the theory behind `new_QEM_method` labeled as `new_QEM_method-5-theory.md`?

The main landing page will link all the sections of the User's Guide such that there is a [workflow diagram](https://mitiq.readthedocs.io/en/latest/contributing_docs.html#adding-workflow-images-to-the-user-guide) for the `new_QEM_method` on this page alongside

> links to any tutorials utilizing `new_QEM_method` in `docs/source/examples`. 

Reviewer: I think this stuff belongs in its own section, not under "Adding files to the user guide". WDYT?

```{code-block}
   # Full Name of New QEM Method

   One line summary of what the technique does

   <start_backticks_code_block>{figure} ../img/new_qem_method_workflow.png
   ---
   width: 700px
   name: new_qem_method_workflow_overview
   ---
   The new_QEM workflow in Mitiq is fully explained in the {doc}`new_QEM_method-4-low-level.md` section.
   <end_backticks_code_block>

   Below you can find sections of the documentation that address the following questions:

   <start_backticks_code_block>{toctree}
   ---
   maxdepth: 1
   ---
   new_QEM_method-1-intro.md
   new_QEM_method-2-use-case.md
   new_QEM_method-3-options.md
   new_QEM_method-4-low-level.md
   new_QEM_method-theory.md
   <end_backticks_code_block>

   A simple tutorial using  `new_QEM_method` can be found <insert_link>

> ```

Reviewer: Does this need to be here? Can we recommend copying it from an existing file?

Author: My intention for this PR was to keep all the information in 1 spot. Copying this from another file adds additional steps.

If you have a notebook you want to add, and want to automatically convert it from the `.ipynb` to `.md`, you can use a great Python command line tool called [jupytext](https://jupytext.readthedocs.io/en/latest/index.html).
To convert from an IPython notebook to markdown file, run `jupytext your_filename.ipynb --to myst` and find the converted file at `your_filename.md`.

> Futher, not only can `jupytext` convert between the formats on demand, but once you install it, you can configure it to manage _both_ a Jupyter and Markdown version of your file, so you don't have to remember to do conversions (for more details, see the `jupytext` docs on [paired notebooks](https://jupytext.readthedocs.io/en/latest/index.html#paired-notebooks)).
> 

Reviewer: ```suggestion
Further, not only can `jupytext` convert between the formats on demand, but once you install it, you can configure it to manage _both_ a Jupyter and Markdown version of your file, so you don't have to remember to do conversions (for more details, see the `jupytext` docs on [paired notebooks](https://jupytext.readthedocs.io/en/latest/index.html#paired-notebooks)).
```

Using the paired notebooks you can continue your development in the notebooks as normal, and just commit to git the markdown serialized version when you want to add to the docs.
You can even add this tool as a [git pre-commit hook](https://jupytext.readthedocs.io/en/latest/using-pre-commit.html) if you want!

```{tip}

> There is a [sample markdown formatted notebook in the `examples` directory](../source/examples/template.md) for you to take a look at as you write your own!

Reviewer: I'm confused by this. It seems both `main` and this branch both work here. Any idea why that is? Is the path with `source` preferred?

Author: Same as my other comment. I had to make sure I specified the full path. The link would not be live otherwise. 

I thin this is because `CONTRIBUTING.md` and `CONTRIBUTING_DOCS.md` files are in separate directories compared to the rest of the files in `docs/`

```

### Adding example files

If you are adding a new tutorial to the [documentation examples](../source/examples/examples.md),

> - add the file as a `.md` file in `docs/source/examples`. A template is available at [this link](../source/examples/template.md).

Reviewer: ```suggestion
- add the file as a `.md` file in `docs/source/examples`. A template is available at [here](../source/examples/template.md).
```
